### ISSUE TITLE
BFV Tensor

### ISSUE THREAD

**Issue by comidan** [OpenMined/TenSEAL#280]
## Question
Hello, is it possible to encrypt or encode a tensor in BFV in order then to be able to make matrix operations such as matmul, dot, etc. in the same way as it is possible with CKKSTensor?

## Further Information
In this [tutorial ](https://github.com/OpenMined/TenSEAL/blob/master/tutorials/Tutorial%202%20-%20Working%20with%20Approximate%20Numbers.ipynb) it's shown the following image. ![image](https://user-images.githubusercontent.com/6382701/114999671-5bfb0480-9ea2-11eb-90b3-f31ae0cde09d.png)
I was wondering if it was being planned, or if it's just possible and I didn't get it, the support for a BFVTensor supporting the various operations already being supported by CKKSTensor.

Thank you for your time!

**Comment by bcebere** [comment]
Hello @comidan 

We offer the [BFVVector](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/cpp/tensors/bfvvector.h), with partial support for some operations. [Tasks are pending](https://github.com/OpenMined/TenSEAL/issues?q=is%3Aissue+is%3Aopen+BFVVector) for adding the API on par with CKKSVector, and any help is welcome.

There is no plan for BFVTensor, but as I said, any contribution is more than appreciated.

Thank you!

**Comment by comidan** [comment]
Hello @bcebere, I'm sorry to reopen the issue. I'm struggling a bit in this task.

The main reason I opened this question was to be able to perform matrix multiplication between BFV encrypted matrices, without having to encrypt every single element of each matrix with BFV scheme but instead representing each matrix as a series of BFV encrypted vector with your BFVVector class.

Do you know any resource I can look into in order to start understanding how could I implement with TenSEAL matrix multiplication using matrices represented as a series of BFV encrypted vectors? And so to contribute with TenSEAL.

Thank you anyway for your time

**Comment by bcebere** [comment]
Hello @comidan 

For matrix multiplication, you can see a reference in the CKKSTensor class. https://github.com/OpenMined/TenSEAL/blob/master/tenseal/cpp/tensors/ckkstensor.cpp#L592

As I said, we only have the BFVVector representation, and we support(or we should support) only matrix x 1D vector operations there. 

A more flexible BFV Tensor would help here, and CKKSTensor is a good reference to do that.

If you want to start working on this new tensor, we can assist.

Thank you!

**Comment by comidan** [comment]
Hello @bcebere
I'm trying to do it, I'm sharing here the initial code.

```python
context = ts.context(ts.SCHEME_TYPE.BFV, 8192, 1032193)

plain = [[(i+j)*3 for i in range(2)] for j in range(2)]

bfv_mat = [ts.bfv_vector(context, plain[i]) for i in range(len(plain))]
bfv_mat1 = [ts.bfv_vector(context, plain[i]) for i in range(len(plain))]

result = []
for i in range(len(bfv_mat)):
    result.append(bfv_mat[i] * bfv_mat1[i])
    #evaluator in theory through sealapi.Evaluator(context) but tenseal context not compatible
    evaluator.add_many(result[i].ciphertext(), result[i].ciphertext())
```
As you can see the problem now is the fact of not having access to a add_many method compatible with __tenseal_cpp.Ciphertext_ but only to _seal::Ciphertext_.
Do you have any suggestions on handling this?

Thank you!

**Comment by bcebere** [comment]
Hello @comidan 

As you can see, the C++ part of BFVVector already implements dot product [here](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/cpp/tensors/bfvvector.cpp#L213). There you can access the evaluator object.

However, the operation is not exposed to the Python bindings.
For that, you need to:
 1, Add the pybind bindings [here](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/binding.cpp#L161). They should be similar to the [CKKSVector dot product bindings](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/binding.cpp#L370)
 2. Wrap the bindings in Python methods [here](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/tensors/bfvvector.py#L74). The methods should be similar to the [CKKSVector ones here](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/tensors/ckksvector.py#L129).

After changing this, you can run `pip install .` and use the dot product via BFVVector API. You can open a PR too with the fix if you wish :)

**Comment by comidan** [comment]
Hello @bcebere 

Thank you very much for the detail guide, and also thank you so that I learnt a bit how is it possible to bind C++ code to Python!

Anyway, I added in [here](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/binding.cpp#L161) the following definition
```python
.def("dot", &BFVVector::dot)
```

And [here](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/tensors/bfvvector.py#L74) the following

```python
    @classmethod
    def _dot(cls, other):
        if isinstance(other, (cls)):
            return other.data
        if not isinstance(other, ts.PlainTensor):
            try:
                other = ts.plain_tensor(other, dtype="int")
            except TypeError:
                raise TypeError(f"can't operate with object of type {type(other)}")
        if len(other.shape) != 1:
            raise ValueError("can only operate with a vector")
        return other.data

    def dot(self, other) -> "BFVVector":
        other = self._dot(other)
        return self._wrap(self.data.dot(other))

    def dot_(self, other) -> "BFVVector":
        other = self._dot(other)
        self.data.dot_(other)
        return self
```

But then when I try to run the following code

```python
import tenseal as ts
import numpy as np
 
 
context = ts.context(ts.SCHEME_TYPE.BFV, 8192, 1032193)
context.generate_galois_keys() #if I don't add this I would get "current context doesn't hold a Galois keys" error
 
plain = [[(i + j)*3 for i in range(2)] for j in range(2)]
 
bfv_mat = [ts.bfv_vector(context, plain[i]) for i in range(len(plain))]
bfv_mat1 = [ts.bfv_vector(context, plain[i]) for i in range(len(plain))]
 
print(bfv_mat[0])
 
result = []
for i in range(len(bfv_mat)):
	for j in range(len(bfv_mat1)):
		result.append(bfv_mat[i].dot_(bfv_mat1[j]))
 
result_clear = np.array([result[i].decrypt() for i in range(len(result))])
print(result_clear)
```

I get as error at the `return self._wrap(self.data.dot(other))` line the following: `RuntimeError: unsupported scheme`

Did I do something wrong? I changed in the code of method _dot the dtype from float to int given that this is for BFV but other than that nothing else.

Thank you again!

**Comment by bcebere** [comment]
Yes, that is correct. I guess the dot product needs more research.

The dot product calls [sum_vector](https://github.com/OpenMined/TenSEAL/blob/master/tenseal/cpp/tensors/utils/utils.cpp#L24).

This calls rotate_vector from SEAL, but this is supported [only for the CKKS scheme](https://github.com/microsoft/SEAL/blob/main/native/src/seal/evaluator.h#L1083).

The SEAL API also provides rotate_rows/columns for BFV, but this needs to be investigated, which one fits the problem.

**Comment by comidan** [comment]
Sorry @bcebere to disturb you again.
Do you think that adding some compatibility to NumPy in order to bypass all these problems is something feasible and "easy" I could do?

I mean in the sense of being able to encrypt instead of a list of integers as it is now if it could be encrypted just one integer and then each of these encrypted integers will be the elements of a NumPy matrix. Now it will be NumPy to efficiently compute dot product and all other operations by having to call the primitives for sum and product of the TenSEAL library when it will interact with its single elements.

Thank you!

**Comment by bcebere** [comment]
TenSEAL handles NumPy objects and converts them to the encrypted space.
There are two flavors: Vector and Tensor. Our main focus is on CKKS since it is more appropriate for machine learning.
The Vector implementation has an optimized memory layout, which means that a ciphertext contains vectors of elements.
The Tensor implementation is more flexible but also more memory-hungry.

The problem is that the Vector representation needs more operations in the encrypted space to compute the dot product. NumPy cannot help here. 

I raised a PR that fixes the dot product for BFV [here](https://github.com/OpenMined/TenSEAL/pull/283).

**Comment by comidan** [comment]
I better understand now, thanks.

Thank you very much @bcebere!

Yes sure, just to have now a working dot product for BFV is what should be needed to make the above code works.
I'll continue to work on that in the following days.

**Comment by comidan** [comment]
Hello @bcebere, today I've been trying to create the BFVTensor.

I was able to successfully compile the library with the new classes, methods and the various C++ - Python bindings.
I was also able to make the Python BFVTensor class pass some tests like sum, sum_batch, power, etc., but I'm stuck at one where I unexpectedly get `ValueError: SEAL: couldn't find context_data from params_id`.

How could I show you the various new classes and code?
Should I open a PR even if not all tests are currently passing for the BFVTensor class operations?

**Comment by bcebere** [comment]
Hello.

Yes, please open a PR.

For C++ tests, you can check out the Github workflows. For running the tests via CMake, you can use https://github.com/OpenMined/TenSEAL/blob/master/.github/workflows/scripts/run_gtest_nix.sh#L3

For Bazel, you can use https://github.com/OpenMined/TenSEAL/blob/master/.github/workflows/bazel_tests.yml#L31

Thank you!

**Comment by comidan** [comment]
> Hello.
> 
> Yes, please open a PR.
> 
> For C++ tests, you can check out the Github workflows. For running the tests via CMake, you can use https://github.com/OpenMined/TenSEAL/blob/master/.github/workflows/scripts/run_gtest_nix.sh#L3
> 
> For Bazel, you can use https://github.com/OpenMined/TenSEAL/blob/master/.github/workflows/bazel_tests.yml#L31
> 
> Thank you!

I added the PR [here](https://github.com/OpenMined/TenSEAL/pull/285).

I don't know why it stops from running the Python tests right at the beginning, on the VM I'm using with Ubuntu 20 they run smooth till the one with the previously described error.

**Comment by comidan** [comment]
Solved with #285, closing the issue.
